### ISSUE TITLE
Support for mixed_float8

### ISSUE THREAD

**Issue by theroggy** [keras-team/keras#23479]
It would be great to be able to use e.g. `keras.config.set_dtype_policy("mixed_float8")` to enable mixed precision inference with float8 to use with e.g. an Nvidia L4 GPU.

As a workaround I tried using the following, which didn't give an error when setting the policy:
``` python
    quantized_policy = keras.dtype_policies.QuantizedFloat8DTypePolicy(
        mode="float8"
    )
    keras.config.set_dtype_policy(quantized_policy)
```

But when running I got the following error anyway: `Layer Functional does not have a `_float8_call` method implemented.`

**Comment by Neilblaze** [comment]
Thanks for the repro @theroggy, that made this easy to track down. I did some digging via Claude, and I found that the error is real, and it turns out to be a bit bigger than float8.

### **What's going on**

When you set a quantized policy globally, every layer picks it up, containers included. `Operation.__call__` sends any layer with a non-None `quantization_mode` through `quantized_call`, which then dispatches to `_float8_call`. Only `Dense` and `EinsumDense` actually implement that. Your `Functional` model object is just the first thing to hit it, but `LayerNormalization`, `Dropout`, `Conv2D` and `MultiHeadAttention` would all raise the same way.

Poking at it a bit, this isn't really a float8 thing. Setting `int8_from_float32` or `int4/-1_from_float32` globally blows up on exactly the same set of layers. `Embedding` dies slightly earlier, inside `build()`. And `model.dtype_policy = <quantized policy>` hits the same wall from the other direction, since it quantizes the sublayers correctly and then leaves the model object itself holding a quantized policy, so the next call still throws.

### **Something that works today**

Skip the global policy and use `model.quantize()` instead. It only touches layers that support the mode and quietly skips the rest.

```python
model = build_model()
model.quantize("float8")
```

### **Worth knowing before you go further with float8**

I'd hold off, honestly. Keras float8 is a QDQ training recipe rather than a post-training inference path. The `Model.quantize` docstring does say so, but it's easy to miss.

Three things fall out of that.

- It doesn't make the model smaller. The kernel stays float32, and float8 adds roughly 12 KB per layer of scale and amax history state on top of it.
- Nothing calibrates the scales outside of training. `inputs_scale` and `kernel_scale` only move through `overwrite_with_gradient` during `optimizer.apply_gradients`. Quantize a trained model, run inference, and they sit at 1.0 forever. Forward passes don't help, not even with `training=True`.
- Whether you get any speed at all comes down to the compiler fusing the quantize-dequantize pair into a real FP8 GEMM. The code is written to match XLA's `gemm_rewrite`, so JAX and TF with XLA. On torch it's just extra work for nothing.

I ran a small MLP through each mode to get a feel for the tradeoff. Relative error against float32, plus total weight bytes.

| mode | weight bytes | median rel err | p99 rel err |
|---|---|---|---|
| `int8` | 0.26x | 1.2% | 0.76 |
| `int4` | 0.15x | 10.7% | 6.25 |
| `float8`, uncalibrated | 1.18x | 6.1% | 3.61 |

That float8 row landing above 1x is partly an artifact of a tiny model, since the scale buffers are a fixed per-layer cost and get lost in the noise on anything real. The shape of it holds though. float8 never gets you below 1x, because the weights stay at full precision.

**What I'd reach for on an L4 instead**

`mixed_bfloat16` as the global policy is well supported on Ada and won't cost you accuracy. Then `model.quantize("int8")` on top if you want the memory back. That one is real PTQ, int8 weights with dynamic per-token activation quantization, and it needs no calibration data at all. L4 has good INT8 tensor cores, so it should be a reasonable fit for what you're after.

**On the name itself**

Small pushback here. `mixed_float16` means compute in float16, so `mixed_float8` reads as compute in float8, which isn't what would happen and isn't really implementable anyway. e4m3 has three mantissa bits, and no backend ships fp8 kernels for softmax or layernorm. The thing you actually want already has a name in the existing scheme, `"float8_from_mixed_bfloat16"`, which is float8 GEMMs with bfloat16 compute and float32 variables.

### **The fix I'd propose**

The dispatch bug seems worth fixing on its own merits. A quantized policy landing on a layer that can't run it should fall through to the normal `call()` rather than raising. The cleanest version I found is to have `Layer.quantization_mode` return `None` when the layer class doesn't implement that mode, while leaving `dtype_policy` alone so it still propagates down to sublayers.

With that in, `set_dtype_policy("float8_from_mixed_bfloat16")` behaves the way you'd expect. `Dense` and `EinsumDense` run float8, including the ones buried inside `MultiHeadAttention`, everything else gets bf16 compute with fp32 variables, and containers stop caring.

I have it prototyped, roughly 30 lines in `layer.py`, and it passes the existing suites unchanged. Two things I'd flag for whoever picks this up. `Layer.stateless_call` has a second dispatch site that needs the same treatment, otherwise JAX `fit` and `predict` still break. And the `dtype_policy` setter needs a guard so it stops calling `quantize()` on layers that never implemented it.

Making float8 genuinely good for inference is a separate and much larger job, a real PTQ calibration pass that fills in the scales from representative data the way GPTQ and AWQ already do. That's probably its own issue.

**Comment by JyotinderSingh** [comment]
Thanks for raising this.

Float8 is already supported through the model quantization API. The recommended workflow is:

```py
keras.config.set_dtype_policy("mixed_bfloat16")
model = build_model(...)
model.quantize("float8")
```

This is also the workflow used in the Keras Float8 training/inference example: https://keras.io/examples/keras_recipes/float8_training_and_inference_with_transformer/

`model.quantize("float8")` applies Float8 quantization only to layers that support it, such as Dense and EinsumDense, while the rest of the model continues to use the source dtype policy.

I would avoid setting `QuantizedFloat8DTypePolicy` through `keras.config.set_dtype_policy()`. A quantized dtype policy is the resulting policy of a layer that has been quantized; it should not be used as the global mechanism for enabling Float8 on a model. Quantization requires layer-specific state and call paths, so it is different from selecting a regular mixed-precision compute policy such as mixed_float16 or mixed_bfloat16.

For the same reason, I don’t think adding a mixed_float8 global dtype policy is the right API. The existing `model.quantize("float8")` API is the intended way to opt a model into Float8.

We can potentially improve the error/message around using quantized policies globally so that users are directed towards `model.quantize("float8")`. We can also consider disabling auto-quantization behaviour in general, when switching dtypes to low precision types.
